### PR TITLE
[WIP] Fer: add SQL order by with priority queue.

### PR DESCRIPTION
**Pull request by dongzl**
**What this PR does**:

**Which issue(s) this PR fixes**:

For #92 .

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/merge/impl/order_by/order_by_value.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	if otherValue == nil {
+		if nullDesc {
+			return 1
+		} else {
+			return -1
+		}
+	}
+	// TODO Deal with case sensitive.
+	// TODO How to compare the interface{}
+	if desc {
```
Comment (jjeffcaii) [inline pkg/merge/impl/order_by/order_by_value.go:107]: How about using `strings.Compare` instead?

File: pkg/merge/impl/order_by/order_priority_queue.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+func (pq *OrderPriorityQueue) Push(x interface{}) {
+	item := x.(*OrderByValue)
+	pq.orderByValues = append(pq.orderByValues, item)
+	pq.update()
+}
+
+func (pq *OrderPriorityQueue) Pop() interface{} {
+	old := *pq
+	n := len(old.orderByValues)
+	item := old.orderByValues[n-1]
```
Comment (jjeffcaii) [inline pkg/merge/impl/order_by/order_priority_queue.go:57]: Add `old.orderByValues[n-1] = nil`, in case of mem leak.

**Review by jjeffcaii — APPROVED** [review]

File: pkg/merge/impl/order_by/order_by_value.go (hunk 1 above)
Comment (dongzl) [reply pkg/merge/impl/order_by/order_by_value.go:107]: > How about using `strings.Compare` instead?

If the value is number type, is it OK?

File: pkg/merge/impl/order_by/order_by_value.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+func compareTo(thisValue, otherValue interface{}, nullDesc, desc, caseSensitive bool) int8 {
+	if thisValue == nil && otherValue == nil {
+		return 0
+	}
+	if thisValue == nil {
+		if nullDesc {
+			return -1
+		} else {
+			return 1
+		}
```
Comment (baerwang) [inline pkg/merge/impl/order_by/order_by_value.go:98]: ignore else 

```go
if nullDesc {
      return -1
}
return 1
```

**Review by baerwang — APPROVED** [review]

File: pkg/merge/impl/order/order_priority_queue.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"container/heap"
+)
+
+type OrderPriorityQueue struct {
+	orderByValues []*OrderByValue
+}
+
+func NewOrderPriorityQueue() OrderPriorityQueue {
```
Comment (AlexStocks) [inline pkg/merge/impl/order/order_priority_queue.go:28]: 月财以前实现过一个优先级队列，你参考下，看看能否复用

**Comment by dongzl** [comment]
code is very old, close this pr and reopen new.

**Pull request closed without merging**